utxo/src/utxo_impl/mod.rs

```diff
@@ -0,0 +1,537 @@
+            };
+
+            let utxo = Utxo {
+                output: output.clone(),
+                // TODO: where do we get the block reward from the transaction?
+                is_block_reward: false,
+                source: source.clone(),
+            };
+
+            self.add_utxo(utxo, &outpoint, overwrite)?;
+        }
+        Ok(())
+    }
+
+    /// Mark the inputs of tx as 'spent'.
+    /// returns a TxUndo if function is a success;
+    /// or an error if the tx's input cannot be spent.
+    pub fn spend_utxos(&mut self, tx: &Transaction, height: BlockHeight) -> Result<TxUndo, Error> {
+        let tx_undo: Result<Vec<Utxo>, Error> = tx
+            .get_inputs()
+            .iter()
+            .map(|tx_in| self.spend_utxo(tx_in.get_outpoint()))
+            .collect();
+
+        self.add_utxos(tx, UtxoSource::BlockChain(height), false)?;
+
+        tx_undo.map(TxUndo::new)
+    }
+
+    /// Adds a utxo entry in the cache.
+    pub fn add_utxo(
+        &mut self,
+        utxo: Utxo,
+        outpoint: &OutPoint,
+        possible_overwrite: bool,
+    ) -> Result<(), Error> {
+        let key = outpoint;
+        let is_fresh = match self.utxos.get(key) {
+            None => {
+                // An insert can be done. This utxo doesn't exist yet, so it's fresh.
+                !possible_overwrite
+            }
+            Some(curr_entry) => {
+                // TODO: update the memory usage
+                // self.memory_usage should be deducted based on this current entry.
+
+                if !possible_overwrite {
+                    if !curr_entry.is_spent() {
+                        // Attempted to overwrite an existing utxo
+                        return Err(Error::OverwritingUtxo);
+                    }
+                    // If the utxo exists in this cache as a 'spent' utxo and is DIRTY, then
+                    // its spentness hasn't been flushed to the parent cache. We're
+                    // re-adding the utxo to this cache now but we can't mark it as FRESH.
+                    // If we mark it FRESH and then spend it before the cache is flushed
+                    // we would remove it from this cache and would never flush spentness
+                    // to the parent cache.
+                    //
+                    // Re-adding a 'spent' utxo can happen in the case of a re-org (the utxo
+                    // is 'spent' when the block adding it is disconnected and then
+                    // re-added when it is also added in a newly connected block).
+                    // if utxo is spent and is not dirty, then it can be marked as fresh.
+                    !curr_entry.is_dirty || curr_entry.is_fresh
+                } else {
+                    // copy from the original entry
+                    curr_entry.is_fresh
+                }
+            }
+        };
+
+        // create a new entry
+        let new_entry = UtxoEntry::new(utxo, is_fresh, true);
```

Comment:
Dirty means it should update the parent; and this case we have to update the parent because the parent doesn't even have it. Fresh means we just created it, so if we spend it in the same round, no need to write it to the parent since the output is created and spent in the same block. Both are optimizations.

And to answer your second concern, ... even better... an enum.

But again, Carla copied bitcoin core... So you're seeing here C++ combined with bitcoin core practices literally translated to rust. 

We'll fix that later. It's not a priority. Once this works properly in tandem with the SpentInputsCache with a working node, we'll be able to clean it up at will.

Reply:
I agree it's not a priority but something to keep in mind